### PR TITLE
FixAll dialog implementation

### PR DESCRIPTION
**Pull request by Therzok**
Current issues:
* IDEXXXX fixes do not work, not sure why, will have to investigate
* There's a minor length check issue when a code fix changes the line count

Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/677579
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/645577

> [Fixes  VSTS #698134](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/698134)

### CODE REVIEW

**Comment by Therzok** [comment]
Removed WIP, noted the minor issues

**Comment by Therzok** [comment]
screen shot 2018-09-21 at 17 05 24

**Comment by Therzok** [comment]
@KirillOsenkov I've noticed a strange thing, sometimes code fixes register the whole document as changed (that's the data coming in from GetTextChanges).

Is it possible we have something missing in our editor implementation or does chunking badly?

Also, how would one translate offsets from the old editor to the new editor? Trying to get exact spans of what code was affected by code fixes

**Comment by vancura** [comment]
Could you please attach light theme screenshot?

**Comment by KirillOsenkov** [comment]
@Therzok when all file changes, could it be the line endings or tabs?

**Comment by mhutch** [comment]
s/occurences/occurrences/

**Comment by mhutch** [comment]
Other than that, it looks great :)

**Comment by Therzok** [comment]
One more issue to fix, you can click the breakpoint margin and breakpoints appear in the editor

**Comment by Therzok** [comment]
screen shot 2018-09-25 at 14 36 01
screen shot 2018-09-25 at 14 37 51

File: main/src/addins/MonoDevelop.Refactoring/MonoDevelop.CodeActions/FixAllPreviewDialog.cs [hunk 1]
```diff
@@ -154,9 +159,24 @@ void UpdateCheckboxDependencies (TreePosition updatedRow)
...
 			UpdateRootFromChild (node, newCheck);
 		}
 
+		void UpdateNextChildrenOffset (TreeNavigator node, CheckBoxState newCheck)
+		{
+			var iter = node.Clone ();
+			var change = node.GetValue (nodeEditor);
+			var diff = change.NewText.Length - change.Span.Length;
+			if (newCheck == CheckBoxState.Off)
+				diff = -diff;
```
Comment (KirillOsenkov) [inline main/src/addins/MonoDevelop.Refactoring/MonoDevelop.CodeActions/FixAllPreviewDialog.cs]: I really think we should start using block statements even for a single-line if body. Remember what happened last time.

Comment (Therzok) [reply main/src/addins/MonoDevelop.Refactoring/MonoDevelop.CodeActions/FixAllPreviewDialog.cs]: True, editorconfig style setting could enforce that.

**Comment by Therzok** [comment]
Ping for review?

File: main/src/addins/MonoDevelop.Refactoring/MonoDevelop.CodeActions/CodeFixMenuService.cs [hunk 2]
```diff
@@ -162,17 +163,40 @@ static CodeFixMenuEntry CreateFixAllMenuEntry (TextEditor editor, FixAllState fi
...
 				// Task.Run here so we don't end up binding the whole document on popping the menu, also there is no cancellation token support
-				var fix = Task.Run (() => provider.GetFixAsync (context));
+				var fix = await Task.Run (() => {
+					var context = fixState.CreateFixAllContext (new RoslynProgressTracker (), token);
+					return provider.GetFixAsync (context);
+				});
+				Microsoft.CodeAnalysis.Text.TextChange[] result = await Task.Run (async () => {
+					var previewOperations = await fix.GetPreviewOperationsAsync (token);
+					return await Runtime.RunInMainThread (() => {
+						var engine = Xwt.Toolkit.CurrentEngine; // NativeEngine
```
[the remark this answers is not in the document]
Comment (Therzok) [reply main/src/addins/MonoDevelop.Refactoring/MonoDevelop.CodeActions/CodeFixMenuService.cs]: NativeEngine does not work yet. Tried it, got this:
```
RROR [2018-09-21 17:44:50Z]: An unhandled exception has occurred. Terminating Visual Studio? False
System.NullReferenceException: Object reference not set to an instance of an object
  at MonoDevelop.Components.Mac.GtkEmbed.UpdateAllocation () [0x00001] in /Users/therzok/Work/md/monodevelop/main/src/core/MonoDevelop.Ide/MonoDevelop.Components/Mac/GtkEmbed.cs:88 
  at MonoDevelop.Components.Mac.GtkEmbed.set_Frame (CoreGraphics.CGRect value) [0x00009] in /Users/therzok/Work/md/monodevelop/main/src/core/MonoDevelop.Ide/MonoDevelop.Components/Mac/GtkEmbed.cs:122 
  at Xwt.Mac.EmbedNativeWidgetBackend.SetNativeView (AppKit.NSView aView) [0x00038] in /Users/therzok/Work/md/monodevelop/main/external/xwt/Xwt.XamMac/Xwt.Mac/EmbedNativeWidgetBackend.cs:51 
  at Xwt.Mac.EmbedNativeWidgetBackend.Initialize () [0x00035] in /Users/therzok/Work/md/monodevelop/main/external/xwt/Xwt.XamMac/Xwt.Mac/EmbedNativeWidgetBackend.cs:27 
  at Xwt.Mac.ViewBackend.Xwt.Backends.IWidgetBackend.Initialize (Xwt.Backends.IWidgetEventSink sink) [0x00008] in /Users/therzok/Work/md/monodevelop/main/external/xwt/Xwt.XamMac/Xwt.Mac/ViewBackend.cs:73 
  at Xwt.Widget+WidgetBackendHost.OnBackendCreated () [0x00001] in /Users/therzok/Work/md/monodevelop/main/external/xwt/Xwt/Xwt/Widget.cs:119 
  at Xwt.Backends.EmbeddedNativeWidget+EmbeddedNativeWidgetBackendHost.OnBackendCreated () [0x00023] in /Users/therzok/Work/md/monodevelop/main/external/xwt/Xwt/Xwt.Backends/IEmbeddedWidgetBackend.cs:67 
  at Xwt.Backends.BackendHost.LoadBackend () [0x00099] in /Users/therzok/Work/md/monodevelop/main/external/xwt/Xwt/Xwt.Backends/BackendHost.cs:170 
  at Xwt.Backends.BackendHost.get_Backend () [0x00001] in /Users/therzok/Work/md/monodevelop/main/external/xwt/Xwt/Xwt.Backends/BackendHost.cs:93 
  at Xwt.Backends.BackendHost`2[T,B].get_Backend () [0x00001] in /Users/therzok/Work/md/monodevelop/main/external/xwt/Xwt/Xwt.Backends/BackendHost.cs:55 
  at Xwt.Widget.get_Backend () [0x00001] in /Users/therzok/Work/md/monodevelop/main/external/xwt/Xwt/Xwt/Widget.cs:376 
  at (wrapper remoting-invoke-with-check) Xwt.Widget.get_Backend()
  at Xwt.Widget.RegisterChild (Xwt.Widget w) [0x00053] in /Users/therzok/Work/md/monodevelop/main/external/xwt/Xwt/Xwt/Widget.cs:1755 
  at Xwt.FrameBox.set_Content (Xwt.Widget value) [0x00022] in /Users/therzok/Work/md/monodevelop/main/external/xwt/Xwt/Xwt/FrameBox.cs:248 
  at Xwt.FrameBox..ctor (Xwt.Widget content) [0x00010] in /Users/therzok/Work/md/monodevelop/main/external/xwt/Xwt/Xwt/FrameBox.cs:137 
  at (wrapper remoting-invoke-with-check) Xwt.FrameBox..ctor(Xwt.Widget)
  at MonoDevelop.CodeActions.FixAllPreviewDialog..ctor (System.String diagnosticId, System.String scopeLabel, Microsoft.CodeAnalysis.CodeFixes.FixAllScope scope, System.Collections.Immutable.ImmutableArray`1[T] operations, MonoDevelop.Ide.Editor.TextEditor baseEditor) [0x0026e] in /Users/therzok/Work/md/monodevelop/main/src/addins/MonoDevelop.Refactoring/MonoDevelop.CodeActions/FixAllPreviewDialog.cs:124 
  at (wrapper remoting-invoke-with-check) MonoDevelop.CodeActions.FixAllPreviewDialog..ctor(string,string,Microsoft.CodeAnalysis.CodeFixes.FixAllScope,System.Collections.Immutable.ImmutableArray`1<Microsoft.CodeAnalysis.CodeActions.CodeActionOperation>,MonoDevelop.Ide.Editor.TextEditor)
  at MonoDevelop.CodeActions.CodeFixMenuService+<>c__DisplayClass4_2+<<CreateFixAllMenuEntry>b__4>d.MoveNext () [0x00012] in /Users/therzok/Work/md/monodevelop/main/src/addins/MonoDevelop.Refactoring/MonoDevelop.CodeActions/CodeFixMenuService.cs:186 
--- End of stack trace from previous location where exception was thrown ---
  at MonoDevelop.Core.Runtime+<>c__42`1+<<RunInMainThread>b__42_0>d[T].MoveNext () [0x00059] in /Users/therzok/Work/md/monodevelop/main/src/core/MonoDevelop.Core/MonoDevelop.Core/Runtime.cs:393 
--- End of stack trace from previous location where exception was thrown ---
  at MonoDevelop.CodeActions.CodeFixMenuService+<>c__DisplayClass4_1+<<CreateFixAllMenuEntry>b__2>d.MoveNext () [0x00102] in /Users/therzok/Work/md/monodevelop/main/src/addins/MonoDevelop.Refactoring/MonoDevelop.CodeActions/CodeFixMenuService.cs:184 
--- End of stack trace from previous location where exception was thrown ---
  at MonoDevelop.CodeActions.CodeFixMenuService+<>c__DisplayClass4_0+<<CreateFixAllMenuEntry>b__0>d.MoveNext () [0x000f2] in /Users/therzok/Work/md/monodevelop/main/src/addins/MonoDevelop.Refactoring/MonoDevelop.CodeActions/CodeFixMenuService.cs:182 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.CompilerServices.AsyncMethodBuilderCore+<>c.<ThrowAsync>b__6_0 (System.Object state) [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono-pullrequest/pr/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/AsyncMethodBuilder.cs:1018 
  at MonoDevelop.Ide.DispatchService+GtkSynchronizationContext+TimeoutProxy.HandlerInternal (System.IntPtr data) [0x00016] in /Users/therzok/Work/md/monodevelop/main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/DispatchService.cs:77 
```

File: main/src/addins/MonoDevelop.Refactoring/MonoDevelop.CodeActions/CodeFixMenuService.cs [hunk 3]
```diff
@@ -162,17 +163,40 @@ static CodeFixMenuEntry CreateFixAllMenuEntry (TextEditor editor, FixAllState fi
...
+						return engine.Invoke (async () => {
+							using (var dialog = new FixAllPreviewDialog (string.Join (", ", fixState.DiagnosticIds), doc.Name, fixState.Scope, previewOperations, editor)) {
```
[the remark this answers is not in the document]
Comment (Therzok) [reply main/src/addins/MonoDevelop.Refactoring/MonoDevelop.CodeActions/CodeFixMenuService.cs]: 👍

**Comment by Therzok** [comment]
@sevoku said the checkbox bug is gtk behaviour. Native toolkit is not an option due to it not working. I can try to add a workaround that detects which cell was clicked to see if it should toggle.

**Comment by slluis** [comment]
That behavior can be changed, like this: https://github.com/mono/monodevelop/blob/document-extensions/main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Components/ProjectSelectorWidget.cs#L64

**Comment by Therzok** [comment]
Merged the Xwt fix.

**Comment by slluis** [comment]
The toggle issue is not fixed in build 739.

**Comment by Therzok** [comment]
Rebased on master with xwt fix.

**Comment by slluis** [comment]
@monojenkins backport release-7.7

**Pull request merged**